src/script_opt/Expr.cc

```diff
@@ -1238,10 +1238,28 @@ ExprPtr EqExpr::Reduce(Reducer* c, StmtPtr& red_stmt) {
     if ( IsHasElementsTest() )
         return BuildHasElementsTest()->Reduce(c, red_stmt);
 
-    if ( GetType()->Tag() == TYPE_BOOL && same_singletons(op1, op2) ) {
-        bool t = Tag() == EXPR_EQ;
-        auto res = with_location_of(make_intrusive<ConstExpr>(val_mgr->Bool(t)), this);
-        return res->Reduce(c, red_stmt);
+    if ( GetType()->Tag() == TYPE_BOOL ) {
+        if ( same_singletons(op1, op2) ) {
+            bool t = Tag() == EXPR_EQ;
+            auto res = with_location_of(make_intrusive<ConstExpr>(val_mgr->Bool(t)), this);
+            return res->Reduce(c, red_stmt);
+        }
+
+        if ( op1->GetType()->Tag() == TYPE_BOOL ) {
```

Comment:
I've merged as is, but a comment would've been nice. The `t = ! t` part is pretty magic, particularly without context what optimization is being employed

Reply:
Thanks. I'll make a note to add that next time around